File: API/CSharp_API/Utility/WebFrameworkUtility.cs
namespace Framework.Utility;

using System.Runtime.InteropServices;
using System.Text.Json.Nodes;
using Framework.Exceptions;

public static partial class WebFrameworkUtility
{
	[LibraryImport(DLLHandler.LIBRARY_NAME)]
	private static partial void deleteWebFrameworkString(IntPtr implementation);

	[LibraryImport(DLLHandler.LIBRARY_NAME)]
	private static partial IntPtr getDataFromString(IntPtr implementation);

	[LibraryImport(DLLHandler.LIBRARY_NAME, StringMarshalling = StringMarshalling.Utf8)]
	private static partial IntPtr createJsonParserFromString(string jsonData, ref IntPtr exception);

	[LibraryImport(DLLHandler.LIBRARY_NAME, StringMarshalling = StringMarshalling.Utf8)]
	private static partial IntPtr getJsonParserParsedData(IntPtr implementation, [MarshalAs(UnmanagedType.Bool)] bool weak, ref IntPtr exception);

	[LibraryImport(DLLHandler.LIBRARY_NAME)]
	private static partial void deleteWebFrameworkJsonParser(IntPtr implementation);

	[UnmanagedFunctionPointer(CallingConvention.Cdecl)]
	private delegate void ProgressCallback(float progress, [MarshalAs(UnmanagedType.LPUTF8Str)] string assetPath, IntPtr data);

	[LibraryImport(DLLHandler.LIBRARY_NAME, StringMarshalling = StringMarshalling.Utf8)]
	private static partial void generateBinaryAssetFile(string directoryPath, string outputPath, ProgressCallback progressCallback, IntPtr data, ref IntPtr exception);

	[LibraryImport(DLLHandler.LIBRARY_NAME, StringMarshalling = StringMarshalling.Utf8)]
	private static partial IntPtr createJwtWithString(IntPtr data, long expirationTimeInMinutes, string jwtSecretVariableName, ref IntPtr exception);

	[LibraryImport(DLLHandler.LIBRARY_NAME)]
	private static partial IntPtr createJwtWithContext(IntPtr data, long expirationTimeInMinutes, IntPtr frameworkInstance, ref IntPtr exception);

	private static string GetStringData(IntPtr stringImplementation)
	{
		string result = Marshal.PtrToStringUTF8(getDataFromString(stringImplementation))!;

		deleteWebFrameworkString(stringImplementation);

		return result;
	}

	public static IList<SqlValue> MakeSqlValues(params object[] args)
	{
		SqlValue[] array = new SqlValue[args.Length];

		for (int i = 0; i < array.Length; i++)
		{
			array[i] = new(args[i]);
		}

		return array;
	}

	public static void GenerateBinaryAssetFile(string directoryPath, string outputPath, Action<float, string, object?>? progressCallback = null, object? data = null)
	{
		IntPtr exception = IntPtr.Zero;
		GCHandle? handle = null;

		if (data != null)
		{
			handle = GCHandle.Alloc(data);
		}

		generateBinaryAssetFile
		(
			directoryPath,
			outputPath,
			(float progress, string assetPath, IntPtr data) =>
			{
				if (progressCallback != null)
				{
					object? temp = null;

					if (data != IntPtr.Zero)
					{
						temp = GCHandle.FromIntPtr(data).Target;
					}

					progressCallback(progress, assetPath, temp);
				}
			},
			handle != null ? GCHandle.ToIntPtr((GCHandle)handle) : IntPtr.Zero,
			ref exception
		);

		handle?.Free();

		if (exception != IntPtr.Zero)
		{
			throw new WebFrameworkException(exception);
		}
	}

	public static string CreateJwt(JsonObject data, TimeSpan expirationTime, string jwtSecretVariableName = "JWT_SECRET")
	{
		IntPtr exception = IntPtr.Zero;
		IntPtr jsonParser = createJsonParserFromString(data.ToJsonString(), ref exception);

		if (exception != IntPtr.Zero)
		{
			throw new WebFrameworkException(exception);
		}

		IntPtr jsonData = getJsonParserParsedData(jsonParser, true, ref exception);

		if (exception != IntPtr.Zero)
		{
			deleteWebFrameworkJsonParser(jsonParser);

			throw new WebFrameworkException(exception);
		}

		IntPtr result = createJwtWithString(jsonData, (long)expirationTime.TotalMinutes, jwtSecretVariableName, ref exception);

		if (exception != IntPtr.Zero)
		{
			deleteWebFrameworkJsonParser(jsonParser);

			throw new WebFrameworkException(exception);
		}

		deleteWebFrameworkJsonParser(jsonParser);

		return GetStringData(result);
	}

	public static string CreateJwt(JsonObject data, TimeSpan expirationTime, WebFramework frameworkInstance)
	{
		IntPtr exception = IntPtr.Zero;
		IntPtr jsonParser = createJsonParserFromString(data.ToJsonString(), ref exception);

		if (exception != IntPtr.Zero)
		{
			throw new WebFrameworkException(exception);
		}

		IntPtr jsonData = getJsonParserParsedData(jsonParser, true, ref exception);

		if (exception != IntPtr.Zero)
		{
			deleteWebFrameworkJsonParser(jsonParser);

			throw new WebFrameworkException(exception);
		}

		IntPtr result = createJwtWithContext(jsonData, (long)expirationTime.TotalMinutes, frameworkInstance.implementation, ref exception);

		if (exception != IntPtr.Zero)
		{
			deleteWebFrameworkJsonParser(jsonParser);

			throw new WebFrameworkException(exception);
		}

		deleteWebFrameworkJsonParser(jsonParser);

		return GetStringData(result);
	}
}

File: API/CC_API/src/utility/web_framework_utility.c
#include "utility/web_framework_utility.h"

web_framework_exception_t wf_generate_uuid(web_framework_string_t* result)
{
	web_framework_exception_t exception = NULL;

	typedef void* (*generateWebFrameworkUUID)(void** exception);

	*result = CALL_WEB_FRAMEWORK_FUNCTION(generateWebFrameworkUUID, &exception);

	return exception;
}

web_framework_exception_t wf_generate_sha256(const char* data, size_t size, web_framework_string_t* result)
{
	web_framework_exception_t exception = NULL;

	typedef void* (*generateSha256)(const char* data, size_t size, void** exception);

	*result = CALL_WEB_FRAMEWORK_FUNCTION(generateSha256, data, size, &exception);

	return exception;
}

web_framework_exception_t wf_generate_binary_asset_file(const char* directory_path, const char* output_path, void(*progress_callback)(float progress, const char* asset_path, void* data), void* data)
{
	web_framework_exception_t exception = NULL;

	typedef void (*generateBinaryAssetFile)(const char* directoryPath, const char* outputPath, void(*progressCallback)(float progress, const char* assetPath, void* data), void* data, void** exception);

	CALL_WEB_FRAMEWORK_FUNCTION(generateBinaryAssetFile, directory_path, output_path, progress_callback, data, &exception);

	return exception;
}

web_framework_exception_t wf_create_jwt_with_string(json_object_t* data, int64_t expiration_time_in_minutes, const char* jwt_secret_variable_name, web_framework_string_t* result)
{
	web_framework_exception_t exception = NULL;

	typedef void* (*createJwtWithString)(void* jsonData, int64_t expirationTimeInMinutes, const char* jwtSecretVariableName, void** exception);

	*result = CALL_WEB_FRAMEWORK_FUNCTION(createJwtWithString, data->implementation, expiration_time_in_minutes, jwt_secret_variable_name ? jwt_secret_variable_name : "JWT_SECRET", &exception);

	return exception;
}

web_framework_exception_t wf_create_jwt_with_context(json_object_t* data, int64_t expiration_time_in_minutes, web_framework_t framework_instance, web_framework_string_t* result)
{
	web_framework_exception_t exception = NULL;

	typedef void* (*createJwtWithContext)(void* jsonData, int64_t expirationTimeInMinutes, void* frameworkInstance, void** exception);

	*result = CALL_WEB_FRAMEWORK_FUNCTION(createJwtWithContext, data->implementation, expiration_time_in_minutes, framework_instance.implementation, &exception);

	return exception;
}
